### PR TITLE
Enhanced printer compatiblity for USB

### PR DESCRIPTION
**Pull request by plokko**
Added auto USB interface and endpoint discovery:
some printer may have the correct interface and endpoint in a differend order, this fix will search all interfaces and endpoint for the correct configuration.

### CODE REVIEW

File: escposprinter/src/main/java/com/dantsu/escposprinter/connection/usb/UsbOutputStream.java [hunk 1]
```diff
@@ -20,7 +21,22 @@
...
     public UsbOutputStream(UsbManager usbManager, UsbDevice usbDevice) throws IOException {
         this.usbInterface = usbDevice.getInterface(0);
-        this.usbEndpoint = this.usbInterface.getEndpoint(1);
+        for(int i=0; i<usbDevice.getInterfaceCount();i++){
+            UsbInterface usbInt = usbDevice.getInterface(i);
+            if(usbInt.getInterfaceClass()== UsbConstants.USB_CLASS_PRINTER){
+                this.usbInterface = usbInt;
+                break;
+            }
+        }
```
Comment (plokko) [inline escposprinter/src/main/java/com/dantsu/escposprinter/connection/usb/UsbOutputStream.java:30]: find the correct interface

File: escposprinter/src/main/java/com/dantsu/escposprinter/connection/usb/UsbOutputStream.java [hunk 2]
```diff
@@ -20,7 +21,22 @@
...
+
+        this.usbEndpoint = this.usbInterface.getEndpoint(0);
+        int endpointCount = this.usbInterface.getEndpointCount();
+        for(int i=0; i<endpointCount;i++){
+            UsbEndpoint endpoint = this.usbInterface.getEndpoint(i);
+            if(endpoint.getType()==UsbConstants.USB_DIR_IN){
+                this.usbEndpoint = endpoint;
+            }
+        }
```
Comment (plokko) [inline escposprinter/src/main/java/com/dantsu/escposprinter/connection/usb/UsbOutputStream.java:39]: finds the correct endpoint (input)

File: escposprinter/src/main/java/com/dantsu/escposprinter/connection/usb/UsbPrintersConnections.java [hunk 3]
```diff
@@ -58,7 +59,13 @@ public static UsbConnection selectFirstConnected(Context context) {
...
 
             if(usbClass == UsbConstants.USB_CLASS_PER_INTERFACE) {
-                usbClass = device.getInterface(0).getInterfaceClass();
+                for(int x=0; x<device.getInterfaceCount();x++){
+                    UsbInterface usbInt = device.getInterface(x);
+                    if(usbInt.getInterfaceClass()== UsbConstants.USB_CLASS_PRINTER){
+                        usbClass = UsbConstants.USB_CLASS_PRINTER;
+                        break;
+                    }
+                }
```
Comment (plokko) [inline escposprinter/src/main/java/com/dantsu/escposprinter/connection/usb/UsbPrintersConnections.java:68]: Fix printer not displaying on list if not standard

**Comment by DantSu** [comment]
Thanks. It's a very helpfull PR !

You forgot break on endpoint loop.
I think it will be better to throwing Exception if interface or endpoint aren't found. Like this : 

```java
    public UsbOutputStream(UsbManager usbManager, UsbDevice usbDevice) throws IOException {
        this.usbInterface = null;
        this.usbEndpoint = null;

        int
                interfacesCount = usbDevice.getInterfaceCount(),
                endpointsCount = this.usbInterface.getEndpointCount();

        for (int i = 0; i < interfacesCount; i++) {
            UsbInterface usbInt = usbDevice.getInterface(i);
            if (usbInt.getInterfaceClass() == UsbConstants.USB_CLASS_PRINTER) {
                this.usbInterface = usbInt;
                break;
            }
        }
        
        if(this.usbInterface == null) {
            throw new IOException("Unable to find USB interface.");
        }
        
        for (int i = 0; i < endpointsCount; i++) {
            UsbEndpoint endpoint = this.usbInterface.getEndpoint(i);
            if (endpoint.getType() == UsbConstants.USB_DIR_IN) {
                this.usbEndpoint = endpoint;
            }
        }

        if(this.usbEndpoint == null) {
            throw new IOException("Unable to find USB endpoint.");
        }
        
        this.usbConnection = usbManager.openDevice(usbDevice);
        
        if(this.usbConnection == null) {
            throw new IOException("Unable to open USB connection.");
        }
    }
```

**Comment by DantSu** [comment]
This is a copy past : 

```java
               for(int x=0; x<device.getInterfaceCount();x++){
                    UsbInterface usbInt = device.getInterface(x);
                    if(usbInt.getInterfaceClass()== UsbConstants.USB_CLASS_PRINTER){
                        usbClass = UsbConstants.USB_CLASS_PRINTER;
                        break;
                    }
                }
```

```java
        for (int i = 0; i < interfacesCount; i++) {
            UsbInterface usbInt = usbDevice.getInterface(i);
            if (usbInt.getInterfaceClass() == UsbConstants.USB_CLASS_PRINTER) {
                this.usbInterface = usbInt;
                break;
            }
        }
```

Something better can be do. Like a method : 

#### UsbConnection.java

```java
    static public UsbInterface findPrinterInterface(UsbDevice usbDevice) {
        if(usbDevice != null) {
            int interfacesCount = usbDevice.getInterfaceCount();
            for (int i = 0; i < interfacesCount; i++) {
                UsbInterface usbInterface = usbDevice.getInterface(i);
                if (usbInterface.getInterfaceClass() == UsbConstants.USB_CLASS_PRINTER) {
                    return usbInterface;
                }
            }
        }
        return null;
    }

    static public UsbEndpoint findEndpointIn(UsbInterface usbInterface) {
        if(usbInterface != null) {
            int endpointsCount = usbInterface.getEndpointCount();
            for (int i = 0; i < endpointsCount; i++) {
                UsbEndpoint endpoint = usbInterface.getEndpoint(i);
                if (endpoint.getType() == UsbConstants.USB_DIR_IN) {
                    return endpoint;
                }
            }
        }
        return null;
    }
```

#### UsbOutputStream.java

```java
    public UsbOutputStream(UsbManager usbManager, UsbDevice usbDevice) throws IOException {
        this.usbInterface = UsbConnection.findPrinterInterface(usbDevice);
        this.usbEndpoint = UsbConnection.findEndpointIn(this.usbInterface);

        if(this.usbInterface == null) {
            throw new IOException("Unable to find USB interface.");
        }

        if(this.usbEndpoint == null) {
            throw new IOException("Unable to find USB endpoint.");
        }

        this.usbConnection = usbManager.openDevice(usbDevice);

        if(this.usbConnection == null) {
            throw new IOException("Unable to open USB connection.");
        }
    }
```

#### UsbPrintersConnections.java

```java
    public UsbConnection[] getList() {
        UsbConnection[] usbConnections = super.getList();
    
        if(usbConnections == null) {
            return null;
        }

        int i = 0;
        UsbConnection[] printersTmp = new UsbConnection[usbConnections.length];
        for (UsbConnection usbConnection : usbConnections) {
            UsbDevice device = usbConnection.getDevice();
            int usbClass = device.getDeviceClass();

            if(usbClass == UsbConstants.USB_CLASS_PER_INTERFACE && UsbConnection.findPrinterInterface(device) != null) {
                usbClass = UsbConstants.USB_CLASS_PRINTER;
            }

            if (usbClass == UsbConstants.USB_CLASS_PRINTER) {
                printersTmp[i++] = new UsbConnection(this.usbManager, device);
            }
        }

        UsbConnection[] usbPrinters = new UsbConnection[i];
        System.arraycopy(printersTmp, 0, usbPrinters, 0, i);
        return usbPrinters;
    }
```

**Comment by plokko** [comment]
> I think it will be better to throwing Exception if interface or endpoint aren't found. Like this :
> 
I was wondering if throwing an exception or select the first interface available;
i've chosen the latter as, in some non-standard printer, it may still work even if it have a wrong interface type (you still have to force it as it won't display on the printer list)

**Comment by DantSu** [comment]
Ok, always exceptions with non conventionnal device ! Classic... ^^
Ok so something like below will be better : 
 
```java
    static public UsbInterface findPrinterInterface(UsbDevice usbDevice) {
        if (usbDevice == null) {
            return null;
        }
        int interfacesCount = usbDevice.getInterfaceCount();
        for (int i = 0; i < interfacesCount; i++) {
            UsbInterface usbInterface = usbDevice.getInterface(i);
            if (usbInterface.getInterfaceClass() == UsbConstants.USB_CLASS_PRINTER) {
                return usbInterface;
            }
        }
        return usbDevice.getInterface(0);
    }

    static public UsbEndpoint findEndpointIn(UsbInterface usbInterface) {
        if (usbInterface == null) {
            return null;
        }

        int endpointsCount = usbInterface.getEndpointCount();
        for (int i = 0; i < endpointsCount; i++) {
            UsbEndpoint endpoint = usbInterface.getEndpoint(i);
            if (endpoint.getType() == UsbConstants.USB_DIR_IN) {
                return endpoint;
            }
        }

        return usbInterface.getEndpoint(1);
    }
```

**Comment by plokko** [comment]
Ok, will leave exception for endpoint as it should be always in.
By the way i already applied all your modifications and fixed a possible nullpointer exception in your code.

**Comment by DantSu** [comment]
perfect !

**Comment by DantSu** [comment]
This is not perfect ! 
`endpoint.getType() == UsbConstants.USB_DIR_IN` is wrong => `endpoint.getDirection() == UsbConstants.USB_DIR_IN`

this is the real condition : `endpoint.getType() == UsbConstants.USB_ENDPOINT_XFER_BULK && endpoint.getDirection() == UsbConstants.USB_DIR_IN`

**Comment by plokko** [comment]
> This is not perfect !
> `endpoint.getType() == UsbConstants.USB_DIR_IN` is wrong => `endpoint.getDirection() == UsbConstants.USB_DIR_IN`
> and is not `USB_DIR_IN `but `USB_DIR_OUT`.
> 
> this is the real condition : `endpoint.getType() == UsbConstants.USB_ENDPOINT_XFER_BULK && endpoint.getDirection() == UsbConstants.USB_DIR_OUT`

Good to hear:
 i found this solution by reverse enginnering the device infos and your library so i'm glad you found out what i missed in the implementation.
I'll test it on a MUNBYN printer (that didn't work as it had multiple interfaces and endpoints on a different order) and will post the results.

**Comment by DantSu** [comment]
Great ! Say me if it works well ! Say me if it doesn't works too 😁

**Comment by plokko** [comment]
Works correctly with MUNBYN ITPP047.

**Pull request merged**